### ISSUE TITLE
"RuntimeError: Could not determine the location of the Python C API library" with v1.0.0

### ISSUE THREAD

**Issue by sshameer** [cropsinsilico/cis_interface#55]
Hi

I installed yggdrasil v1.0.0 from PyPi and ran my YAML file using yggrun and came across the the error **RuntimeError: Could not determine the location of the Python C API library : libpython3.6m.so.**. I came across a similar error with Python 2.7 as well. I do not have this issue with v0.10.0. The full error message is in the file below:

[errorLOgYggv100.txt](https://github.com/cropsinsilico/yggdrasil/files/4231290/errorLOgYggv100.txt)

**Comment by langmm** [comment]
Do you know where the libpython*.so file is on your machine? This will depend on how you installed Python and I might have missed a common installation location.

**Comment by sshameer** [comment]
I do not remember exactly how I installed python (been a while now), but I must have used `sudo apt-get install python3.6`

libpython3.6m.so on my computer is available at 
/snap/gnome-3-28-1804/110/usr/lib/python3.6/config-3.6m-x86_64-linux-gnu/libpython3.6m.so
/snap/gnome-3-28-1804/116/usr/lib/python3.6/config-3.6m-x86_64-linux-gnu/libpython3.6m.so
/usr/lib/x86_64-linux-gnu/libpython3.6m.so
/usr/lib/python3.6/config-3.6m-x86_64-linux-gnu/libpython3.6m.so

**Comment by langmm** [comment]
There are a few things you can try that might give us more information:
1. Run `yggconfig` and send the contents of the file that is printed (and any errors if they occur).
2. Run `ygginfo` and send the output.
3. Run `which python` and `python --version`. It looks like you might have multiple versions of python installed.
4. Locate the python header & static library associated with the executable found by `which` and add those paths to the config file from step 1 manually in the C section using the keys `python_include` and `python_static` respectively.
5. If 4 doesn't work and there aren't any obvious issues with the output from 1, 2, or 3, you could try installing via conda (dev instructions [here](https://cropsinsilico.github.io/yggdrasil/development/general.html#setting-up-a-dev-environment)).

**Comment by sshameer** [comment]
1. Ran `yggconfig`. I received an error message but config file was generated (see yggconfig_Error and yggdrasil_cfg_copy.txt attached
[yggDrasil_debugging.zip](https://github.com/cropsinsilico/yggdrasil/files/4261088/yggDrasil_debugging.zip)
)
2. Ran `ygginfo` and received a similar error (see ygginfo_Error attached)
3. `which python3` returns `/usr/bin/python3` and `python --version` returns `Python 3.6.9`

For 4, I'm not sure what I should be looking for. Do they any specific names? (like python.h maybe?
[yggDrasil_debugging.zip](https://github.com/cropsinsilico/yggdrasil/files/4262181/yggDrasil_debugging.zip)
)

**Comment by langmm** [comment]
What operating system are you on again?

At the minimum I think there is a bug in how the libraries are searched for on Mac that I will work on (namely looking for .so rather than .dylib). But I think there is another issue where the files may be stored by default in a location yggdrasil is not currently searching both on Mac (installed via brew or system default) and linux (installed via apt).

1. Does `which python` (no 3) return the same path? I know my mac has python2 installed as python so I am surprised that `python --version` returns 3.6, but maybe your laptop/OS dosn't have 2.7 installed.
2. If the python you want to use is not the default python (returned by `which python`) then you will need to set the `interpreter` key in the python section of the config file to that path.
3. The output from running `yggconfig` and `ygginfo` look the same and unfortunately don't provide any additional information.
4. On my machine (mac) the native python header (for the `python_include` key in the c section of the config file) is `/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/Python.h` and the library is `/System/Library/Frameworks/Python.framework/Versions/2.7/lib/libpython2.7.dylib` (would be stored under the `python_shared` key in the c section of the config file). Depending on if your version of Python was installed by default or installed via brew/apt it may be in a similar location or not. You can get try the following script to get the header & library locations, but it is not a guarantee.

```
import os, glob
from distutils.sysconfig import get_python_inc, get_python_lib
python_include = os.path.join(get_python_inc(), 'Python.h')
python_shared = glob.glob(os.path.join(os.path.dirname(get_python_lib(True, True)), 'libpython*.*'))[0]
print('include: %s\nshared: %s' % (python_include, python_shared))
```

**Comment by sshameer** [comment]
Hi

Interesting. It is curious though that v0.10.0 works fine for me.

Anyway, 
1) Sorry that was supposed to be `python3 --version`. `which python` return `/usr/bin/python` and `python --version` does return `Python 2.7.17`
4) Thanks, I'll look for those files and let you know how things go

**Comment by sshameer** [comment]
Forgot to mention that I am using Kubuntu 18.04 64-bit

I am afraid I am not doing (4) correctly. My **Python.h** is in "/usr/include/python3.6m/" and **libpython3.6m.so** is in "/usr/lib/x86_64-linux-gnu/". I added them to the config file and yet when I run yggrun, I get the libpython3.6m.so not found error.
![Screenshot_20200227_172032](https://user-images.githubusercontent.com/5559135/75468821-c8a4fe00-5985-11ea-9dfb-4b2e570e0a59.png)

**Comment by langmm** [comment]
So version 0.10 did not compile the C library against the Python library so this wouldn't have been an issues (I added this feature to allow passing Python objects to C to allow use of Python functions). 

1. To use python3 (this is probably the one you want since 2.7 is EOL), you'll need to add `interpreter = /usr/bin/python3` to the python section of your config file.
2. Since libpython3.6m.so is a shared object library, you'll want that to use the key `python_shared` in the config file instead of `python_static`. Sorry, I'm so used to using static libraries since they are available with conda installations of python

**Comment by langmm** [comment]
Oh and you will probably want to run yggconfig before trying yggrun just to ensure the rest of the configuration is done.

**Comment by sshameer** [comment]
I'm afraid yggconfig and yggrun still returns the same error
![Screenshot_20200228_100509](https://user-images.githubusercontent.com/5559135/75539316-f5065c00-5a11-11ea-8ec8-314e70589ae2.png)

The fact that the error has not changed even after I've edited the config file makes me think that yggdrasil must be using a default config file that was generated earlier (probably during installation). If there was one, where would that be?

**Comment by sshameer** [comment]
@langmm , Is there a way to make sure I'm using a particular config file when running yggdrasil?

**Comment by langmm** [comment]
``yggconfig --show-file`` will print the location of the config file that it is reading.

**Comment by langmm** [comment]
So there is a bug where the error you are getting may be raised even when the config variable is set. I have fixed this on the topic/config branch that will be merged in during the next release. You could try installing from that branch for now until I have pushed the release (instructions for a dev setup with install from source is [here](https://cropsinsilico.github.io/yggdrasil/development/general.html#setting-up-a-dev-environment)). If you installed via pypi using pip, you will need to uninstall the version of yggdrasil from pip (`pip uninstall yggdrasil`) before installing from source.

**Comment by langmm** [comment]
This was fixed in PR #56.
